### ISSUE TITLE
Putting and getting a float does not preserve the exact value

### ISSUE THREAD

**Issue by robertnishihara** [amplab/ray#433]
```python
import ray
ray.init(start_ray_local=True, num_workers=0)
x = 0.9
ray.get(ray.put(x))  # prints 0.8999999761581421
```

**Comment by robertnishihara** [comment]
The problem is not in the Python C extensions. For example, defining a simple function

```c++
static PyObject* convert_float(PyObject* self, PyObject* args) {
  PyObject* obj;
  if (!PyArg_ParseTuple(args, "O", &obj)) {
    return NULL;
  }
  double d = PyFloat_AsDouble(obj);
  return PyFloat_FromDouble(d);
}
```

And calling `convert_float` preserves the value.

**Comment by robertnishihara** [comment]
Note the following.

```python
import numpy as np
float(np.float32(0.9))  # prints 0.8999999761581421
```

**Comment by robertnishihara** [comment]
This is addressed by #435.
